### PR TITLE
Model-Level Constraints

### LINKED ISSUE #569: [CT-2221] [BigQuery] Add model-level constraints to get_columns_spec_ddl
Core issue: https://github.com/dbt-labs/dbt-core/issues/6754

BigQuery only supports the not_null constraint

constraint type | support
-- | --
not_null | enforced
primary_key | not supported
foreign_key | not supported
unique | not supported
check | not supported

BigQuery does not support model-level constraints, so no need to render them in the ddl. We'll just need to ensure tests are passing - skipping any model-level constraint tests inherited from the base adapter tests class.  

Constraint docs: https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language

### PR DESCRIPTION
**Pull request by peterallenwebb**
resolves #569 

### Description

Adds BigQuery support for model-level constraints. Core issue: https://github.com/dbt-labs/dbt-core/issues/6754

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-bigquery/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-bigquery/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: dbt/adapters/bigquery/impl.py [hunk 1]
```diff
@@ -907,3 +909,14 @@ def python_submission_helpers(self) -> Dict[str, Type[PythonJobHelper]]:
             "cluster": ClusterDataprocHelper,
             "serverless": ServerlessDataProcHelper,
         }
+
+    @classmethod
+    def render_column_constraint(cls, constraint: ColumnLevelConstraint) -> str:
+        if constraint.type != ConstraintType.not_null:
+            return ""  # Only not null column constraints are supported by BigQuery
+        else:
```
Comment (emmyoop) [inline dbt/adapters/bigquery/impl.py:917]: It also supports primary and foreign keys now.  See https://github.com/dbt-labs/dbt-bigquery/issues/568

Comment (peterallenwebb) [reply dbt/adapters/bigquery/impl.py]: Supported PKs and FKs in last two commits.

File: dbt/adapters/bigquery/impl.py [hunk 2]
```diff
@@ -907,3 +909,26 @@ def python_submission_helpers(self) -> Dict[str, Type[PythonJobHelper]]:
...
+
+    @classmethod
+    def render_column_constraint(cls, constraint: ColumnLevelConstraint) -> str:
+        if constraint.type == ConstraintType.not_null:
+            return super().render_column_constraint(constraint)
+        elif constraint.type == ConstraintType.primary_key:
+            return "primary key not enforced"
+        elif constraint.type == ConstraintType.foreign_key:
+            c = super().render_column_constraint(constraint)
+            return f"{c} not enforced"
```
Comment (VersusFacit) [inline dbt/adapters/bigquery/impl.py:921]: They seem like two distinct semantics; one where you call that constraint and another for which you don't. From an observer's point of view, it feels like there should be separate debugging messages of these two things don't get completed. You are being overcautious?

File: dbt/adapters/bigquery/impl.py [hunk 3]
```diff
@@ -907,3 +909,26 @@ def python_submission_helpers(self) -> Dict[str, Type[PythonJobHelper]]:
...
+        else:
+            return ""
```
Comment (VersusFacit) [inline dbt/adapters/bigquery/impl.py:923]: Any particular reason this method returns an empty string while the method below returns a `None`? What's the contract of the upstream class?

**Review by VersusFacit — COMMENTED** [review]
Looking pretty good, but I wanna understand some of the thinking here better

File: dbt/adapters/bigquery/impl.py [hunk 4]
```diff
@@ -907,3 +909,26 @@ def python_submission_helpers(self) -> Dict[str, Type[PythonJobHelper]]:
...
+
+    @classmethod
+    def render_model_constraint(cls, constraint: ModelLevelConstraint) -> Optional[str]:
+        if (
+            constraint.type == ConstraintType.primary_key
```
[the remark this answers is not in the document]
Comment (peterallenwebb) [reply dbt/adapters/bigquery/impl.py:928]: Yes. These are the only table-level constraints supported by BigQuery. In fact, even these two constraints are very new and not yet in GA.

**Review by McKnight-42 — APPROVED** [review]

**Pull request merged**